File: src/main/java/io/outfoxx/typescriptpoet/SymbolSpecs.kt
/*
 * Copyright 2017 Outfox, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.outfoxx.typescriptpoet


/**
 * Specifies a symbol and its related origin, either via import or implicit/local declaration
 *
 * @param value Value of the symbol
 */
sealed class SymbolSpec(
   open val value: String
) {

  companion object {

    private val fileNamePattern = """(?:[a-zA-Z0-9._\-]+)""".toRegex()
    private val modulePattern = """@?(?:(?:!$fileNamePattern)|(?:$fileNamePattern(?:/$fileNamePattern)*))""".toRegex()
    private val identPattern = """(?:(?:[a-zA-Z][_a-zA-Z0-9]*)|(?:[_a-zA-Z][_a-zA-Z0-9]+))""".toRegex()
    private val importPattern = """($identPattern)?([*@+])($modulePattern)(?:#($identPattern))?""".toRegex()

    /**
     * Parses a symbol reference pattern to create a symbol. The pattern
     * allows the simple definition of all symbol types including any possible
     * import variation. If the spec to parse does not follow the proper format
     * an implicit symbol is created from the unparsed spec.
     *
     * Pattern: `<symbol_name>? <import_type> <module_path> (#<augmented_symbol_name>)?`
     *
     * * symbol_name = `[a-zA-Z0-9._]+`
     *
     *        Any legal compound JS/TS symbol (e.g. symbol._member.member). If no symbol name is
     *        specified then the last component of the module path is used as the symbol name;
     *        allows easy use with libraries that follow normal conventions.
     *
     * * import_type = `@ | * | +`
     *
     *        `@` = Import named symbol from module (e.g. `import { <symbol_name> } from '<module_name>'`)
     *
     *        `*` = Import all symbols from module (e.g. `import * from '<module_name>'`)
     *
     *        `+` = Symbol is declared implicitly via import of the module (e.g. `import '<module_name>'`)
     *
     * * module_path = `!<filename> | <filename>(/<filename)*`
     *
     *        Path name specifying the module. If the module path begins with a `!` then it is considered
     *        to be a file being generated. This ensures the paths are output as relative imports.
     *
     * * augmented_symbol_name = `[a-zA-Z0-9_]+`
     *
     *        Any valid symbol name that represents the symbol that is being augmented. For example,
     *        the import `rxjs/add/observable/from` attaches the `from` method to the `Observable` class.
     *        To import it correctly the spec should be `+rxjs/add/observable/from#Observable`. Adding this
     *        parameter to augmented imports ensures they are output only when the symbol being augmented
     *        is actually used.
     *
     *
     * @param spec Symbol spec to parse.
     * @return Parsed symbol specification
     */
    @JvmStatic
    fun from(spec: String): SymbolSpec {
      val matched = importPattern.matchEntire(spec)
      if (matched != null) {
        val modulePath = matched.groups[3]!!.value
        val type = if (matched.groups[2] != null) {
          matched.groups[2]?.value!!
        }
        else {
          "@"
        }
        val symbolName = matched.groups[1]?.value ?: modulePath.split('/').last().dropWhile { it == '!' }
        val targetName = matched.groups[4]?.value
        return when (type) {
          "*" -> importsAll(symbolName, modulePath)
          "@" -> importsName(symbolName, modulePath)
          "+" -> {
            require(targetName != null) { "Augmenting imports require the name of the augmented symbol" }
            augmented(symbolName, modulePath, targetName!!)
          }
          else -> throw IllegalArgumentException("Invalid type character")
        }
      }

      return implicit(spec)
    }

    /**
     * Creates an import of all the modules exported symbols as a single
     * local named symbol
     *
     * e.g. `import * as Engine from 'templates';`
     *
     * @param localName The local name of the imported symbols
     * @param from The module to import the symbols from
     */
    @JvmStatic
    fun importsAll(localName: String, from: String): SymbolSpec {
      return ImportsAll(localName, from)
    }

    /**
     * Creates an import of a single named symbol from the module's exported
     * symbols.
     *
     * e.g. `import { Engine } from 'templates';`
     *
     * @param exportedName The symbol that is both exported and imported
     * @param from The module the symbol is exported from
     */
    @JvmStatic
    fun importsName(exportedName: String, from: String): SymbolSpec {
      return ImportsName(exportedName, from)
    }

    /**
     * Creates a symbol that is brought in by a whole module import
     * that "augments" an existing symbol.
     *
     * e.g. `import 'rxjs/add/operator/flatMap'`
     *
     * @param symbolName The augmented symbol to be imported
     * @param from The entire import that does the augmentation
     * @param target The symbol that is augmented
     */
    @JvmStatic
    fun augmented(symbolName: String, from: String, target: String): SymbolSpec {
      return Augmented(symbolName, from, target)
    }

    /**
     * An implied symbol that does no tracking of imports
     *
     * @param name The implicit symbol name
     */
    @JvmStatic
    fun implicit(name: String): SymbolSpec {
      return Implicit(name)
    }

  }

  open fun reference(trackedBy: SymbolReferenceTracker?): String {
    trackedBy?.apply { referenced(this@SymbolSpec) }
    return value
  }

  /**
   * Non-imported symbol
   */
  data class Implicit
  internal constructor(
     override val value: String
  ) : SymbolSpec(value) {

    override fun reference(trackedBy: SymbolReferenceTracker?): String {
      return value
    }

  }


  /**
   * Common base class for imported symbols
   */
  abstract class Imported(
     override val value: String,
     open val source: String
  ) : SymbolSpec(value)


  /**
   * Imports a single named symbol from the module's exported
   * symbols.
   *
   * e.g. `import { Engine } from 'templates';`
   */
  data class ImportsName
  internal constructor(
     override val value: String,
     override val source: String
  ) : Imported(value, source)


  /**
   * Imports all of the modules exported symbols as a single
   * named symbol
   *
   * e.g. `import * as Engine from 'templates';`
   */
  data class ImportsAll
  internal constructor(
     override val value: String,
     override val source: String
  ) : Imported(value, source)

  /**
   * A symbol that is brought in by a whole module import
   * that "augments" an existing symbol.
   *
   * e.g. `import 'rxjs/add/operator/flatMap'`
   */
  data class Augmented
  internal constructor(
     override val value: String,
     override val source: String,
     val augmented: String
  ) : Imported(value, source)

}

File: src/SymbolSpecs.ts
import _ from 'lodash';
import Path from 'path';
import { Node } from './Node';

const fileNamePattern = '(?:[a-zA-Z0-9._-]+)';
const modulePattern = `@?(?:(?:${fileNamePattern}(?:/${fileNamePattern})*))`;
const identPattern = `(?:(?:[a-zA-Z][_a-zA-Z0-9.]*)|(?:[_a-zA-Z][_a-zA-Z0-9.]+))`;
export const importType = '[*@+=]';
const importPattern = `^(${identPattern})?(${importType})(${modulePattern})(?:#(${identPattern}))?`;

/**
 * Specifies a symbol and its related origin, either via import or implicit/local declaration.
 *
 * @param value Value of the symbol
 */
export abstract class SymbolSpec extends Node {
  /**
   * Parses a symbol reference pattern to create a symbol. The pattern
   * allows the simple definition of all symbol types including any possible
   * import variation. If the spec to parse does not follow the proper format
   * an implicit symbol is created from the unparsed spec.
   *
   * Pattern: `symbolName? importType modulePath (#<augmentedSymbolName>)?`
   *
   * Where:
   *
   * - `symbolName` is any legal JS/TS symbol. If none, we use the last part of the module path as a guess.
   * - `importType` is one of `@` or `*` or `+`, where:
   *    - `@` is a named import
   *       - `Foo@bar` becomes `import { Foo } from 'bar'`
   *    - `*` is a star import,
   *       - `*Foo` becomes `import * as Foo from 'Foo'`
   *       - `Foo*foo` becomes `import * as Foo from 'foo'`
   *    - `+` is an implicit import
   *       - E.g. `Foo+foo` becomes `import 'foo'`
   * - `modulePath` is a path
   *    - E.g. `<filename>(/<filename)*`
   * - augmentedSymbolName = `[a-zA-Z0-9_]+`
   *
   *        Any valid symbol name that represents the symbol that is being augmented. For example,
   *        the import `rxjs/add/observable/from` attaches the `from` method to the `Observable` class.
   *        To import it correctly the spec should be `+rxjs/add/observable/from#Observable`. Adding this
   *        parameter to augmented imports ensures they are output only when the symbol being augmented
   *        is actually used.
   *
   *
   * @param spec Symbol spec to parse.
   * @return Parsed symbol specification
   */
  public static from(spec: string): SymbolSpec {
    const matched = spec.match(importPattern);
    if (matched != null) {
      const modulePath = matched[3];
      const type = matched[2] || '@';
      const symbolName = matched[1] || _.last(modulePath.split('/')) || '';
      const targetName = matched[4];
      switch (type) {
        case '*':
          return SymbolSpecs.importsAll(symbolName, modulePath);
        case '@':
          return SymbolSpecs.importsName(symbolName, modulePath);
        case '=':
          return SymbolSpecs.importsDefault(symbolName, modulePath);
        case '+':
          return targetName
            ? SymbolSpecs.augmented(symbolName, modulePath, targetName)
            : SymbolSpecs.sideEffect(symbolName, modulePath);
        default:
          throw new Error('Invalid type character');
      }
    }
    return SymbolSpecs.implicit(spec);
  }

  public static fromMaybeString(spec: string | SymbolSpec): SymbolSpec {
    return typeof spec === 'string' ? SymbolSpec.from(spec) : spec;
  }

  /**
   * Defined if the symbol is typically imported from a barrel/etc. path, but is technically defined in another file.
   *
   * We need to know this in case the "this comes from the barrel" type ends up being used in the file where
   * the symbol itself is defined, i.e. we don't need an import in that case.
   */
  public definedIn?: string;

  protected constructor(public value: string) {
    super();
  }

  public toCodeString(): string {
    return this.value;
  }

  public get childNodes(): unknown[] {
    return [];
  }

  public abstract source: string | undefined;
}

export class SymbolSpecs {
  /**
   * Creates an import of all the modules exported symbols as a single
   * local named symbol
   *
   * e.g. `import * as Engine from 'templates';`
   *
   * @param localName The local name of the imported symbols
   * @param from The module to import the symbols from
   */
  public static importsAll(localName: string, from: string): SymbolSpec {
    return new ImportsAll(localName, from);
  }

  /**
   * Creates an import of a single named symbol from the module's exported
   * symbols.
   *
   * e.g. `import { Engine } from 'templates';`
   *
   * @param exportedName The symbol that is both exported and imported
   * @param from The module the symbol is exported from
   */
  public static importsName(exportedName: string, from: string): SymbolSpec {
    return new ImportsName(exportedName, from);
  }

  /**
   * Creates a symbol that is brought in by a whole module import
   * that "augments" an existing symbol.
   *
   * e.g. `import 'rxjs/add/operator/flatMap'`
   *
   * @param symbolName The augmented symbol to be imported
   * @param from The entire import that does the augmentation
   * @param target The symbol that is augmented
   */
  public static augmented(symbolName: string, from: string, target: string): SymbolSpec {
    return new Augmented(symbolName, from, target);
  }

  /**
   * Creates a symbol that is brought in as a side effect of
   * an import.
   *
   * e.g. `import 'mocha'`
   *
   * @param symbolName The symbol to be imported
   * @param from The entire import that does the augmentation
   */
  public static sideEffect(symbolName: string, from: string): SymbolSpec {
    return new SideEffect(symbolName, from);
  }

  /**
   * An implied symbol that does no tracking of imports
   *
   * @param name The implicit symbol name
   */
  public static implicit(name: string): SymbolSpec {
    return new Implicit(name);
  }

  /**
   * Creates an import of a single named symbol from the module's exported
   * default.
   *
   * e.g. `import Engine from 'engine';`
   *
   * @param exportedName The symbol that is both exported and imported
   * @param from The module the symbol is exported from
   */
  public static importsDefault(exportedName: string, from: string): SymbolSpec {
    return new ImportsDefault(exportedName, from);
  }
}

/**
 * Non-imported symbol
 */
export class Implicit extends SymbolSpec {
  constructor(value: string) {
    super(value);
  }

  source = undefined;
}

/** Common base class for imported symbols. */
export abstract class Imported extends SymbolSpec {
  /** The value is the imported symbol, i.e. `BarClass`, and source is the path it comes from. */
  protected constructor(public value: string, public source: string) {
    super(source);
  }
}

/**
 * Imports a single named symbol from the module's exported
 * symbols.
 *
 * E.g.:
 *
 * `import { Engine } from 'templates'` or
 * `import { Engine as Engine2 } from 'templates'`
 */
export class ImportsName extends Imported {
  /**
   * @param value
   * @param source
   * @param sourceValue is the optional original value, i.e if we're renaming the symbol it is `Engine`
   */
  constructor(value: string, source: string, public sourceValue?: string) {
    super(value, source);
  }

  public toImportPiece(): string {
    return this.sourceValue ? `${this.sourceValue} as ${this.value}` : this.value;
  }
}

/**
 * Imports a single named symbol from the module's exported
 * default.
 *
 * e.g. `import Engine from 'engine';`
 */
export class ImportsDefault extends Imported {
  constructor(value: string, source: string) {
    super(value, source);
  }
}

/**
 * Imports all of the modules exported symbols as a single
 * named symbol
 *
 * e.g. `import * as Engine from 'templates';`
 */
export class ImportsAll extends Imported {
  constructor(value: string, source: string) {
    super(value, source);
  }
}

/**
 * A symbol that is brought in by a whole module import
 * that "augments" an existing symbol.
 *
 * e.g. `import 'rxjs/add/operator/flatMap'`
 */
export class Augmented extends Imported {
  constructor(value: string, source: string, public augmented: string) {
    super(value, source);
  }
}

/**
 * A symbol that is brought in as a side effect of an
 * import.
 *
 * E.g. `from("Foo+mocha")` will add `import 'mocha'`
 */
export class SideEffect extends Imported {
  constructor(value: string, source: string) {
    super(value, source);
  }
}

/** Generates the `import ...` lines for the given `imports`. */
export function emitImports(imports: SymbolSpec[], filePath: string): string {
  if (imports.length == 0) {
    return '';
  }

  let result = '';
  const ourModulePath = filePath.replace(/\.[tj]sx?/, '');

  const augmentImports = _.groupBy(filterInstances(imports, Augmented), (a) => a.augmented);

  // Group the imports by source module they're imported from
  const importsByModule = _.groupBy(
    imports.filter((it) => it.source !== undefined && !(it instanceof ImportsName && it.definedIn === ourModulePath)),
    (it) => it.source
  );

  // Output each source module as one line
  Object.entries(importsByModule).forEach(([modulePath, imports]) => {
    // Skip imports from the current module
    if (ourModulePath === modulePath || Path.resolve(ourModulePath) === Path.resolve(modulePath)) {
      return;
    }
    const importPath = maybeRelativePath(ourModulePath, modulePath);

    // Output star imports individually
    filterInstances(imports, ImportsAll).forEach((i) => {
      result += `import * as ${i.value} from '${importPath}';\n`;
      const augments = augmentImports[i.value];
      if (augments) {
        augments.forEach((augment) => (result += `import '${augment.source}';\n`));
      }
    });

    // Output named imports as a group
    const names = unique(filterInstances(imports, ImportsName).map((it) => it.toImportPiece()));
    const def = unique(filterInstances(imports, ImportsDefault).map((it) => it.value));
    if (names.length > 0 || def.length > 0) {
      const namesPart = names.length > 0 ? [`{ ${names.join(', ')} }`] : [];
      const defPart = def.length > 0 ? [def[0]] : [];
      const allNames = [...defPart, ...namesPart];
      result += `import ${allNames.join(', ')} from '${importPath}';\n`;
      [...names, ...def].forEach((name) => {
        const augments = augmentImports[name];
        if (augments) {
          augments.forEach((augment) => (result += `import '${augment.source}';\n`));
        }
      });
    }
  });

  const sideEffectImports = _.groupBy(filterInstances(imports, SideEffect), (a) => a.source);
  Object.keys(sideEffectImports).forEach((it) => (result += `import '${it}';\n`));

  return result;
}

type Constructor<T> = new (...args: any[]) => T;

function filterInstances<T, U>(list: T[], t: Constructor<U>): U[] {
  return (list.filter((e) => e instanceof t) as unknown) as U[];
}

function unique<T>(list: T[]): T[] {
  return [...new Set(list)];
}

export function maybeRelativePath(outputPath: string, importPath: string): string {
  if (!importPath.startsWith('./')) {
    return importPath;
  }
  // Drop the `./` prefix from the outputPath if it exists.
  const basePath = outputPath.replace(/^.\//, '');
  // Ideally we'd use a path library to do all this.
  const numberOfDirs = basePath.split('').filter((l) => l === '/').length;
  if (numberOfDirs === 0) {
    return importPath;
  }
  // Make an array of `..` to get our importPath to the root directory.
  const a: string[] = new Array(numberOfDirs);
  const prefix = a.fill('..', 0, numberOfDirs).join('/');
  return prefix + importPath.substring(1);
}
